xdsl/backend/csl/print_csl.py

```diff
@@ -361,32 +359,53 @@ def print_block(self, body: Block):
                     else:
                         # Use symbol ref name if operand not provided
                         export_val = name
-                    with self._in_block("comptime"):
-                        self.print(f"@export_symbol({export_val}, {q_name});")
+                    with self.descend("comptime") as inner:
+                        inner.print(f"@export_symbol({export_val}, {q_name});")
                 case csl.LayoutOp(body=bdy):
-                    with self._in_block("layout"):
-                        self.print_block(bdy.block)
-                        for name, val in self._symbols_to_export.items():
-                            ty = self.attribute_value_to_str(val[0])
+                    with self.descend("layout") as inner:
+                        inner.print_block(bdy.block)
+                        for name, val in inner._symbols_to_export.items():
+                            ty = inner.attribute_value_to_str(val[0])
                             # If specified, get mutability as true/false from python bool
                             mut = str(val[1]).lower() if val[1] is not None else ""
-                            self.print(f'@export_name("{name}", {ty}, {mut});')
+                            inner.print(f'@export_name("{name}", {ty}, {mut});')
                 case anyop:
                     self.print(f"unknown op {anyop}", prefix="//")
 
-    def descend(self) -> CslPrintContext:
+    @contextmanager
+    def descend(self, block_start: str = ""):
```

Comment:
nit: you can add a type annotation here so that we know the type of the context manager:

```py
from typing import ContextManager
# ...
@contextmanager
def descend(self, block_start: str = "") ->  ContextManager[CslPrintContext]:
    ....
```

Reply:
I'm curious whether this will work, I would have expected it to rather be `-> Iterator[CslPrintContext]`

Reply:
iirc pyright (at least when I tested it) were happy with both, but I find `ContextManager` to be the most correct imo

Reply:
When I try `ContextManager[CslPrintContext]`, pyright produces all sorts of errors. ruff also complains that `typing.ContextManager` is deprecated in favor of `AbstractContextManager`, `AbstractContextManager[CslPrintContext]` causes all the same errors, since it's not compatible with `Generator`.

`Iterator[CslPrintContext]` works, as does `Generator[CslPrintContext, None, None]`, but I think both of these make the function signature more confusing. Personally, I would leave the return type auto-deduced, and rely on the docstring

Reply:
Yeah, the type annotation is for the function passed into the decorator, so `ContextManager` is incorrect